presto-hive/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java

```diff
@@ -0,0 +1,945 @@
+                    continue;
+                }
+                if (metadataFileSystem.isFile(new Path(childPath, PRESTO_SCHEMA_FILE_NAME))) {
+                    childSchemaDirectories.add(childPath);
+                }
+            }
+            return childSchemaDirectories.build();
+        }
+        catch (IOException e) {
+            throw new PrestoException(HIVE_METASTORE_ERROR, e);
+        }
+    }
+
+    private void deleteMetadataDirectory(Path metadataDirectory)
+    {
+        try {
+            Path schemaPath = new Path(metadataDirectory, PRESTO_SCHEMA_FILE_NAME);
+            if (!metadataFileSystem.isFile(schemaPath)) {
+                // if there is no schema file, assume this is not a database, partition or table
+                return;
+            }
+
+            if (!metadataFileSystem.delete(metadataDirectory, true)) {
+                throw new PrestoException(HIVE_METASTORE_ERROR, "Could not delete metadata directory");
+            }
+        }
+        catch (IOException e) {
+            throw new PrestoException(HIVE_METASTORE_ERROR, e);
+        }
+    }
+
+    private <T> Optional<T> readSchemaFile(String type, Path metadataDirectory, JsonCodec<T> codec)
+    {
+        Path schemaPath = new Path(metadataDirectory, PRESTO_SCHEMA_FILE_NAME);
+        return readFile(type + " schema", schemaPath, codec);
+    }
+
+    private <T> Optional<T> readFile(String type, Path path, JsonCodec<T> codec)
+    {
+        try {
+            if (!metadataFileSystem.isFile(path)) {
+                return Optional.empty();
+            }
+
+            try (FSDataInputStream inputStream = metadataFileSystem.open(path)) {
+                byte[] json = ByteStreams.toByteArray(inputStream);
+                return Optional.of(codec.fromJson(json));
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(HIVE_METASTORE_ERROR, "Could not read " + type, e);
+        }
+    }
+
+    private <T> void writeSchemaFile(String type, Path directory, JsonCodec<T> codec, T value, boolean overwrite)
+    {
+        Path schemaPath = new Path(directory, PRESTO_SCHEMA_FILE_NAME);
+        writeFile(type + " schema", schemaPath, codec, value, overwrite);
+    }
+
+    private <T> void writeFile(String type, Path path, JsonCodec<T> codec, T value, boolean overwrite)
+    {
+        try {
+            byte[] json = codec.toJsonBytes(value);
+
+            if (!overwrite) {
+                if (metadataFileSystem.exists(path)) {
+                    throw new PrestoException(HIVE_METASTORE_ERROR, type + " file already exists");
+                }
+            }
+
+            metadataFileSystem.mkdirs(path.getParent());
+
+            // todo implement safer overwrite code
+            try (OutputStream outputStream = metadataFileSystem.create(path, overwrite)) {
+                outputStream.write(json);
+            }
+        }
+        catch (Exception e) {
```

Comment:
This should probably just be `IOException`. Any `RuntimeException` would be a bug